src/java.base/share/classes/java/lang/reflect/Field.java

```diff
@@ -1304,5 +1411,209 @@ public AnnotatedType getAnnotatedType() {
+            }
+        }
+
+        // check if field mutation is enabled for caller module or illegal final field
+        // mutation is allowed
+        var mode = ModuleBootstrap.illegalFinalFieldMutation();
+        if (mode == ModuleBootstrap.IllegalFinalFieldMutation.DENY
+                && !Modules.isFinalMutationEnabled(moduleToCheck(caller))) {
+            throw new IllegalAccessException(callerNotAllowedToMutateMessage(caller, unreflect));
+        }
+    }
+
+    /**
+     * Invoke after mutating a final instance field, or when unreflecting a final instance
+     * field for mutation, to print a warning and record a JFR event.
+     */
+    private void postSetFinal(Class<?> caller, boolean unreflect) {
+        assert isFinalInstanceInNormalClass();
+
+        var mode = ModuleBootstrap.illegalFinalFieldMutation();
+        if (mode == ModuleBootstrap.IllegalFinalFieldMutation.WARN) {
+            // first mutation prints warning
+            Module moduleToCheck = moduleToCheck(caller);
+            if (Modules.tryEnableFinalMutation(moduleToCheck)) {
+                String warningMsg = finalFieldMutationWarning(caller, unreflect);
+                String targetModule = (caller != null && moduleToCheck.isNamed())
+                        ? moduleToCheck.getName()
+                        : "ALL-UNNAMED";
+                VM.initialErr().printf("""
+                        WARNING: %s
+                        WARNING: Use --enable-final-field-mutation=%s to avoid a warning
+                        WARNING: Mutating final fields will be blocked in a future release unless final field mutation is enabled
+                        """, warningMsg, targetModule);
+            }
+        } else if (mode == ModuleBootstrap.IllegalFinalFieldMutation.DEBUG) {
+            // print warning and stack trace
+            var sb = new StringBuilder(finalFieldMutationWarning(caller, unreflect));
+            StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE)
+                    .forEach(sf -> {
+                        sb.append(System.lineSeparator()).append("\tat " + sf);
+                    });
+            VM.initialErr().println(sb);
+        }
+
+        // record JFR event
+        FinalFieldMutationEvent.offer(getDeclaringClass(), getName());
+    }
+
+    /**
+     * Returns the Module to use for access checks with the given caller.
+     */
+    private Module moduleToCheck(Class<?> caller) {
+        if (caller != null) {
+            return caller.getModule();
+        } else {
+            // no java caller, only allowed if field is public in exported package
+            return ClassLoaders.appClassLoader().getUnnamedModule();
+        }
+    }
+
+    /**
+     * Returns the warning message to print when this final field is mutated by
+     * the given possibly-null caller.
+     */
+    private String finalFieldMutationWarning(Class<?> caller, boolean unreflect) {
```

Comment:
It may make sense to have this method return a StringBuilder instance (and use it internally before returning it) as that would cut down on extra String-instance creation.